### PR TITLE
Provide #[derive(MallocSizeOf)] that is actually working

### PR DESCRIPTION
**Pull request by NikVolf**
On top of #290

### CODE REVIEW

File: parity-util-mem/tests/derive.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#[test]
+#[cfg(feature = "std")]
+fn derive_smoky() {
```
Comment (dvdplm) [inline parity-util-mem/tests/derive.rs]: What is "smoky" about this test?

File: parity-util-mem/tests/derive.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	#[derive(MallocSizeOf)]
+	struct Trivia {
+		v: Vec<u8>,
+	}
+
+	let t = Trivia { v: vec![0u8; 1024] };
+
+	assert!(t.malloc_size_of() > 1000);
+}
+
```
Comment (dvdplm) [inline parity-util-mem/tests/derive.rs]: Would be good with a test that exercise the `#[ignore_malloc_size_of = …]` attribute and maybe error messages as well?

File: parity-util-mem/derive/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "parity-util-mem-derive"
+version = "0.1.0"
+authors = ["Parity Technologies"]
+license = "GPL-3.0"
```
Comment (dvdplm) [inline parity-util-mem/derive/Cargo.toml]: [`malloc_size_of_derive`](https://github.com/bholley/malloc_size_of_derive) is MIT/Apach2 – if this is a clean re-implementation I guess it's fine to use GPL but, yeah, I think we should stick to MIT/Apache here.

File: parity-util-mem/Cargo.toml [hunk 4]
```diff
@@ -19,6 +19,7 @@ wee_alloc = { version = "0.4.5", optional = true }
 # The performance penalty is only around 3% on average over our benchmarks.
 mimallocator = { version = "0.1.3", features = ["secure"], optional = true }
 mimalloc-sys = { version = "0.1.6", optional = true }
+parity-util-mem-derive = { path = "derive", version = "0.1" }
```
Comment (dvdplm) [inline parity-util-mem/Cargo.toml]: Need to remove usage of [`malloc_size_of_derive`](https://github.com/paritytech/parity-common/pull/291/files#diff-a6e9f06585ce3cce1d6deea24a49aeb1R12) right?

**Comment by NikVolf** [comment]
@dvdplm addressed

File: parity-util-mem/src/malloc_size.rs [hunk 5]
```diff
@@ -43,7 +43,8 @@
 //!   measured as well as the thing it points to. E.g.
 //!   `<Box<_> as MallocSizeOf>::size_of(field, ops)`.
 
-// This file is patched at commit 5bdea7dc1c80790a852a3fb03edfb2b8fbd403dc DO NOT EDIT.
+//! This is an extended (for own internal needs) version of the Servo internal malloc_size crate.
```
Comment (dvdplm) [inline parity-util-mem/src/malloc_size.rs]:
```suggestion
//! This is an extended version of the Servo internal malloc_size crate.
```

File: parity-util-mem/src/malloc_size.rs [hunk 6]
```diff
@@ -43,7 +43,8 @@
...
+//! We should occasionally track the upstream changes/fixes and reintroduce them here, be they applicable.
```
Comment (dvdplm) [inline parity-util-mem/src/malloc_size.rs]:
```suggestion
//! We should occasionally track the upstream changes/fixes and reintroduce them here, whenever applicable.
```

File: parity-util-mem/tests/derive.rs (hunk 2 above)
Comment (NikVolf) [reply parity-util-mem/tests/derive.rs]: Added `ignore_malloc_size_of` test

**Comment by NikVolf** [comment]
@ordian @niklasad1 Hey guys can you also take a look?

File: parity-util-mem/derive/lib.rs [hunk 7]
```diff
@@ -0,0 +1,65 @@
+//! A crate for deriving the MallocSizeOf trait.
```
Comment (ordian) [inline parity-util-mem/derive/lib.rs]: missing license header

File: parity-util-mem/tests/derive.rs [hunk 8]
```diff
@@ -0,0 +1,47 @@
+use parity_util_mem::{MallocSizeOf, MallocSizeOfExt};
```
Comment (ordian) [inline parity-util-mem/tests/derive.rs]: missing license header

**Comment by NikVolf** [comment]
> AFAIR `#[derive(MallocSizeOf)]` actually worked if we renamed `parity-util-mem` to `memory_size_of` ([example](https://github.com/paritytech/parity-ethereum/blob/b9f9d11929d57114067fb85bcf0bd4112f1d4d69/util/journaldb/Cargo.toml#L13)), but I agree it is a hack.

Yeah, hacky, might bring some issues later

**Comment by cheme** [comment]
Lgtm too, about the way to include external module (parity-util-mem) into the derive macro, I like the current substrate way of doing thing (since we got the derive macro in our code base we can probably use it now): https://github.com/paritytech/substrate/blob/20a9b15cdbed4bf962a4447e8bfb812f766f2fbc/frame/support/procedural/tools/src/lib.rs#L34 (I do not remember to well how it works but it is actually well tested on substrate).
With the code forked, it may be also possible to add a serde like implementation for a specific field with a function (I remember some discussion about it and I am not sure it is worth the work).

**Comment by ordian** [comment]
> With the code forked, it may be also possible to add a serde like implementation for a specific field with a function (I remember some discussion about it and I am not sure it is worth the work).

You mean #255? Yeah, it's possible to implement it now.

**Pull request merged**